### PR TITLE
Joda to JSR 310 Migration

### PR DESCRIPTION
**Pull request by aurambaj**
Note on the Joda dependencies:
- Joda dependencies come from different packages all of which are removed but for the AWS SDK. Amazon has an article on the topic at: https://aws.amazon.com/blogs/developer/java-sdk-bundle/. We could eventually consider a linter to make sure Joda is never used anymore but for now we'll ensure it through code reviews.
- We're adding a "test" dependency on Joda Time to add tests that validate this migration

Add JSR310Migration. JSR310MigrationTest, JSR310MigrationForTesting: those 3 classes are introduced for this migration. They will contain code to make drop in replacement from Joda to JSR310. Some of this code won't make sense to be kept in the long run but in the context of the migration it will help having good visibility on what is changed. The unit tests are here to make sure the basic usage is backward compatible, and in some cases highlight some differences deemed acceptable.

There will be one large commit that does the "drop in" replacement since it is near impossible to have small commit that build/pass tests. The idea is to follow up with smaller refactoring commits
 that will simplify the code where it makes the most sense. But the change history should still be
 relatively easy to track

This PR also adds "threeten-extra" library. It complements the JSR310 support from the JDK and in
 this case is useful to replace Period formatting. Joda Period has not direct equivalent
  in JSR310 and requires "threeten-extra" PeriodDuration.

Beyond drop-in replacement the following major code change are made:

- Remove hardcoded timezone from the surefire plugin for "common", "webapp" and "cli" modules. Running test both locally and with UTC (CI runs UTC) unveiled more errors. Testing in another TZ that UTC make sense since the CLI can be run with a local timezone. This will generally happen when developing/testing locally.

- Must register a DateTimeProvider (@EnableJpaAuditing(dateTimeProviderRef = "zonedDateTimeProvider")). By default, it seems to use LocalDateTime ... but that's a bit odd to not support other types by default. Will check more the doc later. TODO(jean) JSR310

- Configure Jackson serialization/deserialization to use millisecond. To be backward compatible with the Joda time implementation, we need to make sure the timestamp don't include the nano seconds. This is done by SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS, DeserializationFeature.READ_DATE_TIMESTAMPS_AS_NANOSECONDS

  For the record, it helps fix the
TMServiceTest.testExportAssetAsXLIFF, and also fixed the frontend (date would show all wrong)

- Adjust HSQL vs Mysql and now Jackson ser/des different rounding precision

- Remove usage of the type Timestamp, since it was used just in a few place and causing a test to fail. We normalize on using ZonedDateTime everywhere

- Stop using deprecated method isEqualToComparingFieldByField()

- Fix issue with HSQL timezone support & make sure queries are compatible with MySQL. See SQL_FORMATTER and unix_timestamp(:now) in queries

- Renamed PeriodConverter to PeriodDurationConverter

None exhaustive list of changes:
- Removes jackson-datatype-joda from common pom.xml
- Replace jackson datetype to use jsr310 instead of joda
- org.joda.time.DateTime --> java.time.ZonedDateTime (+ import changes)
- org.joda.time.DateTimeZone --> import java.time.ZoneId (+import changes)
- new DateTime() = JSR310Migration.newDateTimeCtorOld --> JSR310Migration.newDateTimeCtor
- new DateTime(0) / new DateTime(EPOCH)--> JSR310Migration.newDateTimeAtEpoch()
- new DateTime(Date) --> JSR310Migration.dateTimeToDate()
- new DateTime(dateTime, dateTimeZone) --> JSR310Migration.newDateTimeCtorWithStringAndDateTimeZone
- new DateTime(dateTimeZone) --> JSR310Migration.newDateTimeCtorWithDateTimeZone- DateTime.now() = JSR310Migration.newDateTimeCtorOld --> JSR310Migration.dateTimeNow
- new DateTime(instant: String|Long) --> JSR310Migration.newDateTimeCtorWithLongAndString
- new DateTime(YMDHM) --> JSR310Migration.newDateTimeCtor(2021, 11, 25, 0, 0);
- DateTime.getMillis() --> JSR310Migration.getMillis()
- dateTime.toDate() --> JSR310Migration.dateTimeToDate()
- dateTime.isAfter(long) --> JSR310Migration.dateTimeIsAfterEpochMillis()
- dateTime.getMillisOfSecond --> JSR310Migration.dateTimeGetMillisOfSecond
- dateTime.withMillisOfSecond --> JSR310Migration.dateTimeWithMillisOfSeconds
- dateTime.withTime(localTime) --> JSR310Migration.dateTimeWithLocalTime
- PeriodFormat.getDefault().print(new Period(start, end)) --> JSR310Migration.toWordBasedDuration() (small difference but accepted since test)
- Period --> DurationPeriod
- new Period(long) --> JSR310Migration.newPeriodCtorWithLong
- new Period(hmsm) --> JSR310Migration.newPeriodCtorWithHMSM
- new LocalTime(str) --> JSR310Migration.newLocalTimeWithString
- new LocalTime(h,m,s) --> JSR310Migration.newLocalTimeWithHMS
- dateTime.getDayOfWeek() --> JSR310Migration.dateTimeGetDayOfWeek
- dateTime.withDayOfWeek(dayOfWeek) --> JSR310Migration.dateTimeWithDayOfWeek
- dateTime.withMillisOfSecond(0).getMillis() --> JSR310Migration.dateTimeWith0MillisAsMillis
- DateTimeZone.forID(zoneId) --> JSR310Migration.dateTimeZoneForId
- Instant.ofEpochSecond(epochSecond).toDateTime() --> JSR310Migration.dateTimeOfEpochSecond
- assertEquals() --> JSR310MigrationForTesting.junitAssertEquals()
- Add Joda time as explicit dependency for testing the migration. Could to be removed later.

### CODE REVIEW

**Comment by aurambaj** [comment]
For the record, https://github.com/pinterest/mojito/pull/58 was the PR for early work on this migration. The commits were not all passing test so it does not make sense to merge them to master, but it can be useful to review some of the change in isolation. It is also missing a few change made in this PR.

**Review by hylstonnb — APPROVED** [review]
leaving a comment, but LGTM thanks

File: webapp/pom.xml [hunk 1]
```diff
@@ -261,30 +261,31 @@
...
-            <artifactId>usertype.core</artifactId>
-            <version>${usertype.core.version}</version>
-        </dependency>
-
-        <dependency>
-            <groupId>joda-time</groupId>
-            <artifactId>joda-time</artifactId>
-            <version>2.10.6</version>
-        </dependency>
+<!--        &lt;!&ndash; Joda Time &ndash;&gt;-->
```
[the remark this answers is not in the document]
Comment (aurambaj) [reply webapp/pom.xml]: looks like I pushed an update (that removed that) while you were commenting but I also pushed some extra files at the same time, cleaning up

File: common/pom.xml [hunk 2]
```diff
@@ -422,7 +436,7 @@
                 <configuration>
                     <!-- Sets the VM argument line used when unit tests are run. -->
                     <argLine>-Dspring.config.additional-location=optional:file://${user.home}/.l10n/config/common/
-                        -Dspring.profiles.active=${user.name},test -Duser.timezone=UTC
+                        -Dspring.profiles.active=${user.name},test
```
Comment (maallen) [inline common/pom.xml:439]: Was it intentional to drop the setting of the user timezone?

**Review by maallen — APPROVED** [review]
LGTM, single comment around the removal on setting the timezone during tests

**Pull request merged**